content/userdefinedfunctions.md

```diff
@@ -0,0 +1,189 @@
+slug: /
+---
+
+User Defined functions provide a lightweight way to implement custom behaviour outside of what Hypi offers. The functions are executed on the same server where the query gets processed.
+
+Currently, you may use two programming language for the user defined functions.
+
++ Groovy: The entire Groovy syntax is available on the hypi platform. Groovy is general a general programming language. Adding custom behaviour using Groovy becomes quick, easy and performant.
+
++ Velocity: The velocity template language is in place to output text. This text may execute dynamic behaviour. E.g., customizing the body of an email, substituting receiver’s name from a variable etc.
+
+> Let’s check how to implement user defined functions. 
+
+We will start with a simple schema as follows: 
+```java
+type Query {
+    inlineGroovyFunction(a: String, b: Int, c: Boolean):Json @tan(type:Groovy, inline: """
+        def map = new java.util.LinkedHashMap()
+        map.put("a", a)
+        map.put("b", b)
+        map.put("c", c)
+        return map
+    """)
+    inlineVelocityFunction(a: String, b: Int, c: Boolean):
+    String @tan(type:Velocity, inline: "$a,$b,$c")
+}
+```
+
+The schema has two user defined functions. `inlineGroovyFunction` and `inlineVelocityFunction` You may define the user defined functions either inline or through a script. `@tan` directive is used to add the functions. 
+
+`inlineGroovyFunction` function puts value of variables a,b and c in a LinkedHashMap and the map is returned in the form of JSON data. `inlineVelocityFunction` uses velocity template to return the formatted string output of variables a,b, and c. You may execute the user defined function as any other GraphQL query.
+
+```java
+{
+  inlineGroovyFunction(a:"val1", b:23, c: true)
+}
+OR
+{
+  inlineVelocityFunction(a:"val1", b:23, c: true)
+}
+
+```
+Result:
+```json
+{
+  "data": {
+    "inlineGroovyFunction": {
+      "a": "val1",
+      "b": 23,
+      "c": true
+    }
+  }
+}
+OR
+{
+  "data": {
+    "inlineVelocityFunction": "val1,23,true"
+  }
+}
+```
+You may pass the variable parameters to the function as well. Just like this:
+```java
+query inlineVelocityFunction($a: String, $b: Int, $c: Boolean){
+  inlineVelocityFunction(a: $a, b: $b, c: $c)
+}
+```
+As you can see in `inlineGroovyFunction`, Java classes are also available to use inline. You may also use a simple Groovy like version as follows.
+```java
+inlineGroovyFunction(a: String, b: Int, c: Boolean):Json @tan(type:Groovy, inline: """
+     return [
+     a: a,
+     b: b,
+     c: c
+     ]
+ """)
+```
+
+#### Function ‘gql’
+
+The function 'gql' is available to execute inside the user defined function. Using this function you may perform mutation or query operations inside the user defined function. The result depends upon the type of operation. If you perform an ‘upsert’ mutation, it will return id, createdBy etc. (Just like normal mutation) If you perform a ‘find’ query, it will return a list.
```

Comment:
One thing missed here is that the `gql` function accepts two parameters. 
1. `query`: String - the GraphQL query to execute
2. values: Map - a map containing the set of variables used in `query`
None of these examples use GraphQL variables, but for more complex cases they will be needed. e.g.

In your example below, `"$a"` would fail if the variable `a` contained a `"` because it'd break the GraphQL syntax.
Instead, what you'd do is this:
```java
  SetBookInfo(a: String, b: Float, c: Int):Json @tan(type:Groovy, inline: """
    return gql('''
    mutation CreateBook($a: String, $b: Float, $c: Int){
        upsert(
            values: { 
                Book: [
                    { 
                        title: "$a",
                        price: $b,
                        authorid: $c
                    }
                ] 
            }
        ) {
            id
        }
    }''', ['a':a, 'b': b, 'c': c])
  """)
```
I made a few key changes here:
1. I replaced the `\"""` with just `'''` - by using three `'` in Groovy, we have the same effect as `"""` BUT no string interpolation, that means that `$a` is not evaluated as a Groovy variable and remains as part of the string.
2. I added GraphQL variables `CreateBook($a: String, $b: Float, $c: Int)` 
2. I added variable values ` ['a':a, 'b': b, 'c': c]` - now, the variables `$a`,`$b` and `$c` are GraphQL variables not Groovy variables. That means a string with `"` won't cause the query to fail.

Reply:
ok.. it worked. 

Should I change get function as well?

Reply:
Done